### ISSUE TITLE
Concurrency and Transaction

### ISSUE THREAD

**Issue by xialvjun** [juanluispaz/ts-sql-query#7]
```ts
import { Table } from "../Table";
import { SqliteConnection } from "../connections/SqliteConnection";
import { BetterSqlite3QueryRunner } from "../queryRunners/BetterSqlite3QueryRunner";
import * as betterSqlite3 from 'better-sqlite3'

class DBConection extends SqliteConnection<'DBConnection'> {
}

const tCompany = new class TCompany extends Table<DBConection, 'TCompany'> {
    id = this.autogeneratedPrimaryKey('id', 'int');
    name = this.column('name', 'string');
    constructor() {
        super('company'); // table name in the database
    }
}()

const db = betterSqlite3(':memory:')
const global_conn = new DBConection(new BetterSqlite3QueryRunner(db))

koaApp.use(async (ctx, next) => {
  // await global_conn.beginTransaction(); // I can not write code like this, it may cause concurrency bug. I must write like:
  const scope_conn = new DBConection(new BetterSqlite3QueryRunner(db));
  scope_conn.beginTransaction();
  xxxx;
});
```

Why not:
```ts
const global_conn = new DBConection(new BetterSqlite3QueryRunner(db))

koaApp.use(async (ctx, next) => {
  const t = await global_conn.beginTransaction(); // t is of interface DBConection
  t.selectFrom();
  t.commit();
  t.rollback();
});
```

**Comment by xialvjun** [comment]
It seems even `PgPoolQueryRunner ` doesn't support concurrent transaction too.

```ts
const global_pool = new Pool(xxx);
const global_connection = new DBConection(new PgPoolQueryRunner(pool));

koa.use(mount('/asfds', async (ctx, next) => {
  // this sql may run in the bottom transaction because 'executeBeginTransaction' set 'this.currentQueryRunner = await this.createQueryRunner()'
  await global_connection.update(xx).xxx; 
}));

koaApp.use(async (ctx, next) => {
  await global_connection.executeBeginTransaction();  
});
```

[`this.currentQueryRunner = await this.createQueryRunner()`](https://github.com/juanluispaz/ts-sql-query/blob/master/src/queryRunners/AbstractPoolQueryRunner.ts#L153)

**Comment by juanluispaz** [comment]
Interesting.

One important thing, I don't expect the ts-sql-query Connection or QueryRunner to be reused due to some internal database-dependant connector restrictions. When you use Koa, Express or another HTTP server, I expect you create one DBConnection per request. I'm going to clarify this point in the documentation.

But, still the point of concurrent usage of Sqlite. Concurrent Sqlite can be tricky, especially if you combine it with async operations in the middle of a transaction.

In the case of better-sqlite3, if you don't mix with async operations, you can perform all operations synchronously. But, ts-sql-query is designed to be used in an async way.

Can you send me a code of how you will do without ts-sql-query?

If you don't mix async operation with your database operations, I can create for you a way to execute BetterSqlite3QueryRunner in a synchronous way using 'synchronous-promise' or similar. The database calls will continue having an async api, but they will be executed in a synchronous way. Could this work for you?

If you mix your database calls with async operations, maybe you need to use 'better-sqlite-pool'.

Let me know your feedback and how you will use it.

**Comment by juanluispaz** [comment]
Be careful; you are creating a single PostgreSQL connection and sharing it between both processes. You must use a dedicated connection per process. If not, both execution will be combined in one with unpredictable behaviour. 

Use the pg Pool class instead of the Client class, and use PgPoolQueryRunner instead of PgQueryRunner. PgPoolQueryRunner will do the work to extract the connection from the pool needed to execute the queries in the context of a DBConection instance.

**Comment by juanluispaz** [comment]
I started a PostgreSQL in a docker, and I executed:

```sql
create table company (
    id serial primary key,
    name varchar(100) not null
);
```

Then I executed your code, and at the end, there were only 4 rows on the table; instead of the 50 that you expect. The problem is you have a single database connection. Let me explain what happened, adding some comments in the output of your code:

```
h1 transaction <-------- Global transaction started
h2 transaction <-------- Nested transaction started
h1 transaction o
h1 insert <-------------- First insert in the nested transaction
h2 transaction o
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 insert
h2 insert
h1 Error: h1 goto rollback <------------ Nested transaction rollback
                           <------------ All previous insert are reverted
    at handler_1 (/Users/juanluispazrojas/Proyectos/test/src/test.ts:37:15)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at async Promise.all (index 0)
    at receive_request (/Users/juanluispazrojas/Proyectos/test/src/test.ts:23:3)
h2 insert <--------------- First insert in the global transaction
h2 insert <--------------- Second
h2 insert <--------------- Thrid
h2 insert <--------------- Fourth/Last insert in the global transaction
h2 commit. <------------ Four rows committed to the database
over
```

That is why it ended with 4 rows in the table; the other ones were inserted in the nested transaction rolled back.

**Lest try to fix it**:

I cleaned up the table executing in the database
```sql
truncate company;
```

I modified your code a little bit to use a connection pool:

```typescript
import { Pool, Client } from 'pg';
import { PgPoolQueryRunner } from 'ts-sql-query/queryRunners/PgPoolQueryRunner'; // I changed this line
import { PostgreSqlConnection } from 'ts-sql-query/connections/PostgreSqlConnection';
import { Table } from 'ts-sql-query/Table';
import _ from 'lodash';

class DBConection extends PostgreSqlConnection<'DBConnection'> {}

const tCompany = new (class TCompany extends Table<DBConection, 'TCompany'> {
  id = this.autogeneratedPrimaryKey('id', 'int');
  name = this.column('name', 'string');
  constructor() {
    super('company'); // table name in the database
  }
})();

const client = new Pool({ // I changed this line
  connectionString: 'postgresql://postgres:mysecretpassword@localhost:5432/postgres'
});

async function receive_request() {
  // await client.connect(); // this is optional, it is not required, the connection is stablished automatically
  await Promise.all([handler_1(100), handler_2(50)]);
  console.log('over');
  process.exit();
}
receive_request();

async function handler_1(n: number) {
  const conn = new DBConection(new PgPoolQueryRunner(client)); // I changed this line
  console.log('h1 transaction');
  await conn.beginTransaction();
  console.log('h1 transaction o');
  try {
    for (const i of _.range(n)) {
      if (i > 45) {
        throw new Error('h1 goto rollback');
      }
      console.log('h1 insert');
      await conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
    }
    await conn.commit();
    console.log('h1 commit');
  } catch (error) {
    console.log('h1', error);
    await conn.rollback();
  }
}

async function handler_2(n: number) {
  const conn = new DBConection(new PgPoolQueryRunner(client)); // I changed this line
  console.log('h2 transaction');
  await conn.beginTransaction();
  console.log('h2 transaction o');
  try {
    for (const i of _.range(n)) {
      console.log('h2 insert');
      await conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
    }
    await conn.commit();
    console.log('h2 commit');
  } catch (error) {
    console.log('h2', error);
    await conn.rollback();
  }
}
```

Then I executed, and, in the end, I got 50 rows in the table as you expected.

**Comment by xialvjun** [comment]
Yeah, I write that on purpose to show API like this will make people easy to write wrong code.

In node project, I used to have a global db object and use it across different request, like:

```ts
const pool = new Pool();
app.get('/abc', async ctx => {
  ctx.body = await pool.xxxx();
});
app.get('/def', async ctx => {
  pool.transaction(t => {
    ctx.body = await t.xxxx();
  });
});
```

In `ts-sql-query`, I already have a global object `const client = new Pool({})`, but I have to to create `DBConnection and QueryRunner` for every request with that gloabl object `client`, it feels somewhat weird.

Well, let it pass.

For `better-sqlite3`, I have code:

```ts
import { ExecutableInsert } from 'ts-sql-query/expressions/insert';
import { ExecutableSelect } from 'ts-sql-query/expressions/select';
import { ExecutableUpdate } from 'ts-sql-query/expressions/update';
import { ExecutableDelete } from 'ts-sql-query/expressions/delete';

type ExecutableSelectResult<E extends ExecutableSelect<any, any, any>> = E extends ExecutableSelect<any, infer R, any> ? R : never;
export function run<Q extends ExecutableSelect<any, any, any>>(query: Q): ExecutableSelectResult<Q>[];
export function run<Q extends ExecutableInsert<any> | ExecutableUpdate<any> | ExecutableDelete<any>>(query: Q): betterSqlite3.RunResult;
export function run(query: any) {
  if (!!query.executeSelectMany) {
    return sqlite3.prepare(query.query()).all();
  }
  return sqlite3.prepare(query.query()).run();
};
```

It use `ts-sql-query` just as a `sql-builder` and `type-provider`, so sqls can be executed synchronously just as in `better-sqlite3`. So concurrent transaction, well, not concurrent, but since it's sync, we can do not worry concurrent transaction.


I think pattern like this is good:
> `ts-sql-query` just act as a `sql-builder` and `type-provider`, let those database driver do their things. Or you can decouple `ts-sql-query` to `builder` and `runner`.

**Comment by juanluispaz** [comment]
When you call `pool.transaction(...)` the pool class create all the required objects for you under the hood.

I chose to create a new DBConnection object every time because it is what (more or less) the database libraries do. Additionally, it allows making complex customisations otherwise really hard to implement. For example: in the next release, you will have the possibility of making the insensitive string comparisons (add I added the possibility to the order by as well) to be case insensitive and accent insensitive using collations. Case insensitive and accent insensitive is very important when you work with languages that use accents like Spanish. But, the collation can be different per user language (that in practice means different per request); you will be able to specify the collation (if you want to use it) when you construct the connection object. The readme in the GitHub repository is already updated with this new functionality.

You can get the query and params and execute by yourself, but you will lose the result adaptation of the returned types and the validations. Some databases return values that are not what are you expecting in the result type. Sqlite and Postgres are easy ones; Oracle (by example) is very complicated. All this logic is implemented in the query runner.

You have to specify the query runner to the connection because you can have different strategies. For example, this one allows you to write in the console what happens with the database:

```typescript
const connection = new DBConection(new ConsoleLogQueryRunner(new PgPoolQueryRunner(client)));
```

Or, for testing propose I open you the possibility (if you want) to test the functionality without call the database using a mock query runner (correctness of the query is validated by ts-sql-query during compiling time):

```typescript
const connection = new DBConection(new MockQueryRunner(
    (type, query, params, index) => {
        // verify your queries here
    }
));
```

Regarding the unit test, depending on your strategy, you don't want to have a global connection object that you call directly. Instead of that, you build a middleware in Koa/Express that set the database in the context object and, for logic take the connection from the context. This approach allows you to make complex unit tests.

I would like you don't depend explicitly on intermediate query representation objects. They are very complex objects that can change to implement new functionality.

If you put all the database operations inside a method your code can look like:

```typescript
app.get('/def', async ctx => {
  const result = await executeOnDB(async conn => {
    await conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
    await conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
    return null;
  });
  // write the request output
});
```

Or even simpler if you perform only one operation:

```typescript
app.get('/def', async ctx => {
  const result = executeOnDB(conn => 
    conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
  );
  // write the request output
});
```

The `executeOnDB` function can be something like this:

```typescript
async function executeOnDB<T>(fn: (conn: DBConnection) => Promise<T>) {
    const conn = (null as any as DBConnection);
    await conn.beginTransaction();
    let result;
    try {
        result = await fn(conn);
    } catch (e) {
        try {
            conn.rollback();
        } catch {
            // Ignore internal error in favour of the external one
        }
        throw e;
    }
    await conn.commit();
    return result;
}
```

I think the best approach for you regarding better-sqlite3 is to have the possibility to use synchronous-promise; I'm going to add this feature for you. This will allows you to use all the advantages offered by ts-sql-query.

The previous approach will work; all the code executed inside the executeOnDB will be executed synchronously even when you program it async. If you need the executeOnDB function returns synchronously (only for better-sqlite3), but I think you don't need it, your code can look like this:

```typescript
app.get('/def', async ctx => {
  const result = executeOnDBSync(async conn => {
    await conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
    await conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
    return null;
  });
  // write the request output
});
```

Or even simpler if you perform only one operation:

```typescript
app.get('/def', async ctx => {
  const result = executeOnDBSync(conn =>
    conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
  );
  // write the request output
});
```

The `executeOnDBSync` function can be something like this:

```typescript
/**
 * This function expects the internal function retunr a syncronous promise.
 * This function unwrap the syncronos promise in a syncronos way returning the result.
 */
function executeOnDBSync<T>(fn: (conn: DBConnection) => Promise<T>) {
    const conn = new DBConection(new PgPoolQueryRunner(pool));
    let returned = false
    let result: any
    let error: any
    conn.beginTransaction().then(() => {
        fn(conn).then(r => {
            result = r
            returned = true
            conn.commit().catch(e => {
                error = e
            })
        }, e => {
            conn.rollback();
            error = e
        })
    }, e => {
        error = e
    })

    if (!returned) {
        throw new Error('You performed an real async operation, not a database operation, inside the function dedicated to call the databse')
    }
    if (error) {
        throw error
    }
    return result
}
```

Let me know if the synchronous-promise will work for you.

**Comment by juanluispaz** [comment]
Let me explain my synchronous-promise proposal for use with better-sqlite3

I think the best approach for you regarding better-sqlite3 is to have the possibility to use synchronous-promise; I'm going to add this feature for you. This will allows you to use all the advantages offered by ts-sql-query.

The approach of have a `executeOnDB` function adapted for synchronous calls will work; all the code executed inside the executeOnDB will be executed synchronously even when you program it async. The important part is you cannot use the async/await syntax. You will need to use a complementary `sync` function that transforms the promises in sync code. For example, your code can look like this:

```typescript
app.get('/def', async ctx => {
  const result = executeOnDB(conn => {
    const id1 = sync(conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert());
    const id2 = sync(conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert());
    return { id1, id2 }
  });
  // write the request output
});
```

Or even simpler if you perform only one operation (the sync call will be made automatically):

```typescript
app.get('/def', async ctx => {
  const result = executeOnDB(conn =>
    conn
        .insertInto(tCompany)
        .values({ name: Math.random().toString(36) })
        .executeInsert();
  );
  // write the request output
});
```

The required code can be something like this:

```typescript
import * as betterSqlite3 from 'better-sqlite3'
import { SynchronousPromise } from "synchronous-promise";

const db = betterSqlite3(':memory:')

/**
 * This function unwrap the syncronous promise in a syncronous way returning the result.
 */
function sync<T>(promise: Promise<T>): T {
    let returned = false;
    let result: any;
    let error: any;
    promise.then(r => {
        result = r;
    }, e => {
        error = e;
    })

    if (!returned) {
        throw new Error('You performed an real async operation, not a database operation, inside the function dedicated to call the databse');
    }
    if (error) {
        throw error;
    }
    return result;
}

/**
 * This function expects the internal function return a syncronous promise or a result
 * This function unwrap the syncronos promise in a syncronos way returning the result.
 */
function executeOnDB<T>(fn: (conn: DBConnection) => Promise<T>): T
function executeOnDB<T>(fn: (conn: DBConnection) => T): T
function executeOnDB<T>(fn: (conn: DBConnection) => any): T {
    const conn = new DBConection(new BetterSqlite3QueryRunner(db, { promise: SynchronousPromise }));
    sync(conn.beginTransaction());
    let result;
    try {
        const fnResult = fn(conn);
        if (fn instanceof SynchronousPromise) {
            result = sync(fnResult);
        } else if (fnResult instanceof Promise) {
            throw new Error('You performed an real async operation, not a database operation, inside the function dedicated to call the databse');
        } else {
            result = fnResult;
        }
    } catch (e) {
        try {
            sync(conn.rollback());
        } catch {
            // Ignore internal error in favour of the external one
        }
        throw e;
    }
    sync(conn.commit());
    return result;
}
```

Let me know if the synchronous-promise will work for you.

**Comment by xialvjun** [comment]
Well, even though I still want just a `sql builder and type provider`(we needn't to hide the database driver api, people can choose whichever they like), `synchronous-promise` and `sync` indeed solved my question(`sync` is just like `await` for `synchronous-promise`, good idea).

**Comment by juanluispaz** [comment]
Synchronous queries implemented in ts-sql-query 1.1.0. See the [Synchronous query runners](https://github.com/juanluispaz/ts-sql-query#synchronous-query-runners) section in the documentation and the [example](https://github.com/juanluispaz/ts-sql-query/blob/master/src/examples/BetterSqlite3SyncronousExample.ts).
